cpp/src/groupby/sort/group_nth_element.cu

```diff
@@ -0,0 +1,190 @@
+  std::enable_if_t<is_fixed_width<T>() && !include_nulls, std::unique_ptr<column>>
+  operator()(column_view const &values,
+             rmm::device_vector<size_type> const& group_labels,
+             rmm::device_vector<size_type> const &group_offsets,
+             size_type num_groups, size_type n,
+             rmm::mr::device_memory_resource *mr, cudaStream_t stream)
+  {
+    using ResultType = cudf::experimental::detail::target_type_t<
+        T, experimental::aggregation::Kind::NTH_ELEMENT>;
+    std::unique_ptr<column> result = make_fixed_width_column(
+                                data_type(type_to_id<ResultType>()), num_groups,
+                                mask_state::UNALLOCATED, stream, mr);
+    if (num_groups == 0) {
+      return result;
+    }
+    auto group_sizes = group_count(values, group_labels, num_groups, mr, stream);
+
+    auto values_view = column_device_view::create(values, stream);
+    auto values_element = thrust::make_transform_iterator(
+      thrust::counting_iterator<size_type>(0),
+      [col=*values_view] __device__(auto i) { return col.element<T>(i);});
+    auto bitmask_iterator = thrust::make_transform_iterator(
+      experimental::detail::make_validity_iterator(*values_view),
+      [] __device__ (auto b) { return static_cast<size_type>(b); });
+    rmm::device_vector<size_type> intra_group_index(values.size());
+    auto exec = rmm::exec_policy(stream)->on(stream);
+    //intra group index for valids only.
+    thrust::exclusive_scan_by_key(exec,
+                          group_labels.begin(),
+                          group_labels.end(),
+                          bitmask_iterator,
+                          intra_group_index.begin()
+                          );
+   //gather the valid index == n
+   thrust::scatter_if(exec, values_element, values_element+values.size(), 
+                 group_labels.begin(), //map
+                 thrust::make_counting_iterator<size_type>(0),
+                 result->mutable_view().begin<T>(),
+                 [n, bitmask_iterator, intra_group_index=intra_group_index.begin()]
+                 __device__ (auto i)  -> bool {
+                 return (bitmask_iterator[i] && intra_group_index[i]==n);
+                 });
+
+   // null if n>=group_size.
+   rmm::device_buffer result_bitmask;
+   size_type result_null_count;
+
+   std::tie(result_bitmask, result_null_count) = experimental::detail::valid_if(
+       group_sizes->view().begin<size_type>(), group_sizes->view().end<size_type>(),
+       [n] __device__(auto s) { return (n < 0) ? s >= (-n) : s > n; });
+   if (result_null_count)
+     result->set_null_mask(std::move(result_bitmask), result_null_count);
+   return result;
+  }
+
+  template <typename T, typename... Args>
+  std::enable_if_t<!is_fixed_width<T>(), std::unique_ptr<column> >
+  operator()(Args&&... args) {
+    //TODO for non-fixed types.
+    CUDF_FAIL("Only fixed width types are supported in nth aggregation");
```

Comment:
This doesn't have to be. You can make this work with string if you compute the global idx  first and use a gather.

Reply:
There are groups for which `idx>group_size` and result should be null. will gather work in those cases?

Reply:
replaced with `detail::gather`.